nvflare/fuel/f3/streaming/file_downloader.py

```diff
@@ -14,198 +14,183 @@
 
+        if not (os.path.isfile(file_name) and os.path.exists(file_name)):
+            raise ValueError(f"file {file_name} does not exist or is not a valid file")
 
-class _ChunkProducer(Producer):
+        self.size = os.path.getsize(file_name)
 
-    def __init__(self, chunk_size=None):
-        Producer.__init__(self)
         if not chunk_size:
             chunk_size = DEFAULT_CHUNK_SIZE
 
         check_positive_int("chunk_size", chunk_size)
+        if file_downloaded_cb:
+            check_callable("file_downloaded_cb", file_downloaded_cb)
         self.chunk_size = chunk_size
+        self.file_downloaded_cb = file_downloaded_cb
+        self.cb_kwargs = cb_kwargs
+        self.logger = get_obj_logger(self)
 
-    def produce(self, ref_id: str, obj, state: dict, requester: str) -> (str, Any, dict):
-        assert isinstance(obj, _File)
+    def produce(self, state: dict, requester: str) -> Tuple[str, Any, dict]:
         received_bytes = 0
         if state:
             received_bytes = state.get(_StateKey.RECEIVED_BYTES, 0)
 
         if not isinstance(received_bytes, int) or received_bytes < 0:
             self.logger.error(f"bad {_StateKey.RECEIVED_BYTES} {received_bytes} from {requester}")
-            return ProduceRC.ERROR, None, None
+            return ProduceRC.ERROR, None, {}
 
-        if received_bytes >= obj.size:
+        if received_bytes >= self.size:
             # already done
-            return ProduceRC.EOF, None, None
+            return ProduceRC.EOF, None, {}
 
-        num_bytes_to_send = min(self.chunk_size, obj.size - received_bytes)
-        with open(obj.name, "rb") as f:
+        num_bytes_to_send = min(self.chunk_size, self.size - received_bytes)
+        with open(self.name, "rb") as f:
             f.seek(received_bytes)
             chunk = f.read(num_bytes_to_send)
 
         self.logger.debug(f"{received_bytes=}; sending {len(chunk)} bytes")
         return ProduceRC.OK, chunk, {_StateKey.RECEIVED_BYTES: received_bytes + len(chunk)}
 
+    def downloaded_to_one(self, to_site: str, status: str):
+        if self.file_downloaded_cb:
+            self.file_downloaded_cb(to_site, status, self.name, **self.cb_kwargs)
 
-class FileDownloader(ObjDownloader):
+    def downloaded_to_all(self):
+        if self.file_downloaded_cb:
+            self.file_downloaded_cb("", "", self.name, **self.cb_kwargs)
 
-    @classmethod
-    def new_transaction(
-        cls,
-        cell: Cell,
-        timeout: float,
-        timeout_cb,
-        **cb_kwargs,
-    ):
-        """Create a new file download transaction.
 
-        Args:
-            cell: the cell for communication with recipients
-            timeout: timeout for the transaction
-            timeout_cb: CB to be called when the transaction is timed out
-            **cb_kwargs: args to be passed to the CB
-
-        Returns: transaction id
-
-        The timeout_cb must follow this signature:
+def add_file(
```

Comment:
minor styling issue, should we move add_file and download_file after the _ChunkConsumer?